Impl. redis active gophers

jacob-pascual opened the pull request:
This PR re-adds support for the Presence commands. 

We need to send redis commands (`ZREM`) every time a websocket client disconnects, to mark them as inactive for that client. 

Before, we were naively just using the redis connection pooling library, but there are way too many websocket events to do that; there are roughly ~200 websocket disconnects per second at peak, which would mean that every single disconnect would connect to redis, issue ZREM, and then disconnect. Lots of connections would time out and it was just bad.

This PR makes it so that there is a fixed-at-runtime number of gophers (redis consumers) that issue commands on behalf of websockets to Redis for those presence commands. That number is 8 currently, but is configurable (`redis_activity_consumers`). Websockets forward commands (`RedisCommand`s) to a queue process (`RedisConsumerPendingList`). That pending list is a linked list of pending redis commands. The queue process dispatches them to a number of workers that each perform the command and report back. 

One concern is that the rate of websocket disconnections (about 200/sec) could be faster than the consumers can dispatch to Redis; if it takes more than 40ms on average from the moment the websocket disconnects to the moment the result returns, then the redis command queue will grow without bound. I have added a StatsD metric for the queue length (`incus.pendingactivityredislen`), and it would be prudent to add another metric for how long each command takes from start to finish. 

Also, nothing stops us from using these for any arbitrary redis commands that websockets do.

jacob-pascual committed:
Impl. redis active gophers

briankassouf commented on redis_consumer.go:
```diff
@@ -0,0 +1,37 @@
+			log.Println("Dequeued one command in consumer")
+		}
+
+		conn, success := r.pool.Get()
```
Should you put this conn back into the poll?

jacob-pascual replied:
! good catch

jacob-pascual committed:
fix: Add configuration var for the connection pool size

briankassouf commented on redis_store_test.go:
```diff
@@ -233,6 +245,9 @@ func TestUserPresence(t *testing.T) {
 	store.MarkInactive("bazbar", "sock1")
 
+	// Give it time to propagate to Redis...
+	time.Sleep(1 * time.Second)
```
We could build this in a way it waits for the command to finish on redis so we don't have to sleep, eh?

briankassouf replied:
You could push a command and a channel to write back on to the slice

jacob-pascual replied:
yeah, I'll refactor it so all the async commands return a channel that I can pull from to block.

jacob-pascual committed:
fix: Return the conn to the pool after consuming

jacob-pascual committed:
fix: Make redis consumer async; remove brittle sleeps in tests

briankassouf commented:
:+1:

lospro7 commented on redis_consumer_pending.go:
```diff
@@ -0,0 +1,93 @@
+			if pendingLength != 0 {
+				front = r.pending.Front()
+				r.pending.Remove(front)
+				r.lock.Unlock()
```
should this be `r.pendingCond.L`?

jacob-pascual replied:
this is a bit confusing. there are two locks (one for the `Cond` and one for the list itself). 

The `Cond` is a mechanism for waiting until the list has an element, i.e. all of the consumer goroutines except one will attempt to acquire it, and whichever succeeds will be able to dequeue from the list of pending commands. I think `cond.Signal()` unlocks the `Cond`'s lock without the caller necessarily taking it, so that it "wakes up" exactly one goroutine. 

The `r.lock` (perhaps this should be renamed to `r.listMutex`) protects against concurrent modifications by all of the consumers _and_ the master goroutine that actually adds to the list.

jacob-pascual replied:
Before, I had each of the consumers polling on the queue via `select` but this is more efficient; all of the goroutines sleep until there are pending commands!

lospro7 commented on redis_consumer_pending.go:
```diff
@@ -0,0 +1,93 @@
+				r.lock.Unlock()
+				break
+			} else {
+				r.lock.Unlock()
```
should this be `r.pendingCond.L`?

jacob-pascual committed:
refactor: Rename redis consumer / queue stuff

RedisConsumerPending -> RedisQueue
    "ConsumeForever" => "ReceiveForever"
    "ForwardForever" => "DispatchForever"
RedisConsumer        -> RedisQueueConsumer

lospro7 commented:
:+1:
